target/linux/ath79/dts/ar9342_ubnt_nanobeam-ac.dts

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: GPL-2.0
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar9342_ubnt_wa.dtsi"
+
+/ {
+	compatible = "ubnt,nanobeam-ac", "ubnt,wa", "qca,ar9342";
+	model = "Ubiquiti NanoBeam AC (WA)";
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		rssi0 {
+			label = "ubnt:blue:rssi0";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		rssi1 {
+			label = "ubnt:blue:rssi1";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		rssi2 {
+			label = "ubnt:blue:rssi2";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		rssi3 {
+			label = "ubnt:blue:rssi3";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <4>;
+	phy4: ethernet-phy@4 {
+		phy-mode = "rgmii";
+		reg = <4>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	/* default for ar934x, except for 1000M and 10M */
+	pll-data = <0x06000000 0x00000101 0x00001313>;
+
+	mtd-mac-address = <&eeprom 0x0>;
+
+	phy-mode = "rgmii";
+	phy-handle = <&phy4>;
+
+	gmac-config {
+		device = <&gmac>;
+		rxd-delay = <3>;
```

Comment:
@TobleMiner Are you 100% sure about this delay values? Did you checked the content of the `0x18070000` reg under AirOS or is this just copy&paste from other DTS file?

Reply:
> I have very little knowledge about what this delay value does

@cybermaus https://en.wikipedia.org/wiki/Media-independent_interface#Reduced_gigabit_media-independent_interface

Reply:
I do not own a NanoBeam myself. I do only happen to have built ubnt_wa, Nanostation AC and Nanostation AC loco support. Those values are guesstimations based on my experience with other Ubiquiti wa devices.

@cybermaus It would be nice if you could conduct fhe following test:

Configure a bridge across WLAN and the ethernet port. Plug the ethernet port into a device with sufficient CPU power to run iperf3 with at least 1 Gbit/s throughput.
Start an iperf3 server (iperf3 -s) on that device.

Now connect a device with at least 2x2 MIMO 802.11ac support to the NanoBeams wireless. This device should again be capable of 1 Gbit/s iperf throughput.

Now run 1 minute long iperf tests from this device in both directions:

* tx: `iperf3 -c <server ip> -t 60`
* rx: `iperf3 -c <server ip> -t 60 -R`

Please post the results.

That should allow us to determine if set RGMII delay values are acceptable.

Reply:
Ahh, crap. I've forgot, that one can't easily use io/devmem2 on airOS (one has to find working toolchain first) and since you've explained, that you don't own this device, it's quite troublesome to verify all the stuff.

Reply:
@cybermaus  If you don't want to waste your time with the iperf3 setup/testing, it might be faster to just download my [devmem2-airos](http://ynezz.true.cz/temp/devmem2-airos) which was working for me on XM devices with airOS 6.1.7 and just run `devmem2-airos 0x18070000` under airOS! (not under OpenWrt it probably won't run anyway) and provide the output here.

Reply:
IMHO it does not matter what Ubiquiti configures. Just as @cybermaus already stated the ethernet port was somewhat flaky with the proprietary firmware but seems to be stable on OpenWRT. Thus sticking to the configuration provided by Ubiquiti might not even be the best of ideas and a proper throughput test is required.

Reply:
Not to worry, I will test with the iperf3 ... later.
It's interesting enough for me to want to test. But I have a day-job first.

Before, it was flaky even when no users and no traffic. (no guests in winterttime which is why I noted the issues, because syslog was empty other then this eth0 message). Now it is not doing that. So either OpenWRT does it better (or OpenWRT is so much worse it does not even discover the eth0 port is down for a second)

But before I re-deploy the device as active spare, I want to see it do the full 450Mpbs it should be capable off. Or a reasonable approximation of it. Or rather, test the eth0 stand alone, and see it approach 1Gbps. 

Especially because, if I use this, I need to change all my NanoBeam AC's. 
Unfortunately, AirMax AC is not compatible with regular 802.11ac. So either they are all AirMax, or are all OpenWRT

Reply:
Alas, I do not have a 1Gb capable passive POE handy. They are all still at the campsite.
I did test with 100Mbps, and iperf3 showed good results, see below, so at least the delay value is not very wrong.

Also, just so you know: this is unrelated to the eth0 problems going away: when I had the OEM firmware problems with the eth0 port, I tried both 1Gb and also 100Mb, tried another cable and tried another switch. Problems with eth0 persisted. So I am still quite positive OpenWRT behaves better on my previously flaky device.

Connecting to host 192.168.190.28, port 5201
[  5] local 192.168.190.49 port 3370 connected to 192.168.190.28 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  11.5 MBytes  96.2 Mbits/sec
[  5]   1.00-2.01   sec  11.2 MBytes  94.1 Mbits/sec
[  5]   2.01-3.01   sec  11.4 MBytes  95.2 Mbits/sec
[  5]   3.01-4.00   sec  11.2 MBytes  94.6 Mbits/sec
[  5]   4.00-5.01   sec  11.2 MBytes  94.3 Mbits/sec
[  5]   5.01-6.00   sec  11.1 MBytes  93.7 Mbits/sec
[  5]   6.00-7.00   sec  11.4 MBytes  95.5 Mbits/sec
[  5]   7.00-8.01   sec  11.2 MBytes  93.8 Mbits/sec
[  5]   8.01-9.01   sec  11.4 MBytes  95.5 Mbits/sec
[  5]   9.01-10.00  sec  11.2 MBytes  94.7 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   113 MBytes  94.8 Mbits/sec                  sender
[  5]   0.00-10.07  sec   113 MBytes  94.1 Mbits/sec                  receiver

iperf Done.

Reply:
Update: hold on, remembered to test both directions:

Connecting to host 192.168.190.49, port 5201
[  5] local 192.168.190.28 port 42696 connected to 192.168.190.49 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  11.2 MBytes  93.7 Mbits/sec    0   39.7 KBytes
[  5]   1.00-2.01   sec  11.3 MBytes  94.8 Mbits/sec   11   42.5 KBytes
[  5]   2.01-3.02   sec  7.42 MBytes  61.7 Mbits/sec    0   49.6 KBytes
[  5]   3.02-4.09   sec  5.20 MBytes  40.5 Mbits/sec    0   49.6 KBytes
[  5]   4.09-5.00   sec  9.52 MBytes  87.9 Mbits/sec    0   52.5 KBytes
[  5]   5.00-6.01   sec  11.4 MBytes  94.8 Mbits/sec    0   52.5 KBytes
[  5]   6.01-7.01   sec  11.4 MBytes  94.8 Mbits/sec    0   52.5 KBytes
[  5]   7.01-8.02   sec  7.53 MBytes  62.8 Mbits/sec    0   59.6 KBytes
[  5]   8.02-9.00   sec  5.00 MBytes  42.7 Mbits/sec    0   63.8 KBytes
[  5]   9.00-10.00  sec  8.52 MBytes  71.4 Mbits/sec    0   68.1 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  88.5 MBytes  74.2 Mbits/sec   11             sender
[  5]   0.00-10.01  sec  88.4 MBytes  74.0 Mbits/sec                  receiver

It seems -s on PC and -c on UBNT is slow.
Need to double check that from another device.

Reply:
@cybermaus Hey, no worrys :smile:. I'm not in a hurry and your job of course comes first.

When testing with iperf3 just bear in mind that the NanoBeam has a pretty weak CPU. Thus testing with iperf3 on the device itself might yield pretty terrible results (~100 Mbit/s or so) even when testing directly via the ethernet port. My images come with htop installed so it should be pretty easy to see when you are CPU bottlenecked.

Reply:
@cybermaus Nice, thanks. At least for 100 Mbit/s it is looking pretty good. It would be nice to see a test on Gigabit though once you have a Gigabit POE injector available.

Reply:
Ok, some mixed news.

The good news: at 100Mb, the iperf3 tests are all good, in both directions. You just need to properly set this up, with no Luci connected and also no htop (top is good enough) and all non-essential processes stopped: 95Mb consistently in both directions, with CPU barely going above 4%

Less good news: I will not have availability on a 1Gb passive POE for a few months, as they are all at the campsite, so I will not repeat this test with 1Gb anytime soon. 

Also less good, but not at all relevant to you or this thread: My flaky eth0 is back. It turns out that if I ping outward (from device to PC, or while connected to Wifi, pinging my router) it can run for hours without missing a single ping. That, combined with the lack of dmesg messages, made me think all was well. But if I ping inward (from PC to device) I miss a ping every so often. Its just not reported in dmesg.

Anyway, let me know if I need to test any new build, like with new kernel versions.

Reply:
Well, I am too triggered. Ordered another POE, should be here soon.

Reply:
So, I got the Gigagbit passive POE and the results are not very good: A consistent 120Mbps in one direction, and 270Mpbs in the other direction. I was hoping to get at least to the same 450Mbps that the NanoBeam WiFi should be capable of.

```
root@OpenWrt:~# iperf3 -c 192.168.1.222 -V  -b 500M -R
iperf 3.6
Linux OpenWrt 4.14.91 #0 Tue Jan 8 14:27:57 2019 mips
Control connection MSS 1452
Time: Tue, 08 Jan 2019 22:06:19 GMT
Connecting to host 192.168.1.222, port 5201
      Cookie: 7etnayryeykpcv5oqxgagprra4aamrmzkgmo
      TCP MSS: 1452 (default)
[  5] local 192.168.1.1 port 56018 connected to 192.168.1.222 port 5201
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  15.0 MBytes   125 Mbits/sec    0   34.0 KBytes
[  5]   1.00-2.00   sec  15.0 MBytes   126 Mbits/sec    0   53.9 KBytes
[  5]   2.00-3.01   sec  15.0 MBytes   125 Mbits/sec    0   78.0 KBytes
[  5]   3.01-4.01   sec  14.9 MBytes   125 Mbits/sec    0   92.2 KBytes
[  5]   4.01-5.00   sec  14.9 MBytes   125 Mbits/sec    0   97.8 KBytes
[  5]   5.00-6.01   sec  15.0 MBytes   125 Mbits/sec    0    111 KBytes
[  5]   6.01-7.00   sec  15.0 MBytes   126 Mbits/sec    0    123 KBytes
[  5]   7.00-8.01   sec  15.0 MBytes   126 Mbits/sec    0    123 KBytes
[  5]   8.01-9.01   sec  15.1 MBytes   126 Mbits/sec    0    187 KBytes
[  5]   9.01-10.01  sec  14.8 MBytes   124 Mbits/sec    0    187 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.01  sec   150 MBytes   125 Mbits/sec    0             sender
[  5]   0.00-10.04  sec   150 MBytes   125 Mbits/sec                  receiver
CPU Utilization: local/sender 96.7% (0.5%u/96.1%s), remote/receiver 19.6% (7.2%u/12.4%s)
snd_tcp_congestion cubic

iperf Done.
root@OpenWrt:~# iperf3 -c 192.168.1.222 -V  -b 500M -R
iperf 3.6
Linux OpenWrt 4.14.91 #0 Tue Jan 8 14:27:57 2019 mips
Control connection MSS 1452
Time: Tue, 08 Jan 2019 22:06:33 GMT
Connecting to host 192.168.1.222, port 5201
Reverse mode, remote host 192.168.1.222 is sending
      Cookie: 357lx3xg7r2zkxv3bwnyr3j7kg2vymmiohbq
      TCP MSS: 1452 (default)
[  5] local 192.168.1.1 port 56022 connected to 192.168.1.222 port 5201
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec  31.9 MBytes   266 Mbits/sec
[  5]   1.00-2.00   sec  31.6 MBytes   266 Mbits/sec
[  5]   2.00-3.00   sec  31.9 MBytes   267 Mbits/sec
[  5]   3.00-4.00   sec  32.0 MBytes   268 Mbits/sec
[  5]   4.00-5.00   sec  32.1 MBytes   270 Mbits/sec
[  5]   5.00-6.00   sec  32.2 MBytes   270 Mbits/sec
[  5]   6.00-7.00   sec  32.2 MBytes   271 Mbits/sec
[  5]   7.00-8.00   sec  32.4 MBytes   271 Mbits/sec
[  5]   8.00-9.00   sec  32.1 MBytes   270 Mbits/sec
[  5]   9.00-10.00  sec  32.1 MBytes   270 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.03  sec   321 MBytes   268 Mbits/sec                  sender
[  5]   0.00-10.00  sec   321 MBytes   269 Mbits/sec                  receiver
CPU Utilization: local/receiver 75.7% (0.7%u/75.1%s), remote/sender 10.9% (1.5%u/9.4%s)
rcv_tcp_congestion cubic

iperf Done.
root@OpenWrt:~#
```

Reply:
@cybermaus So now you can play with the `rxd-delay = <3>;` and `rxdv-delay = <3>;` values and find better performing combination. Values can be in range 0-3, where 0 is probably same as removing/not adding that property.

Reply:
@cybermaus It would still be nice to try to get values as used in airOS as described in https://github.com/openwrt/openwrt/pull/1733#discussion_r246577676

Reply:
Why yes, but how? This is where my lack of experience comes in. I am just a test monkey.

Short from setting up a full build environment and building myself (and I have not done that for OpenWRT since White Russia days) I have no idea how to play with these values. 

Anyone knows of a small utility that can set/get these registers directly? Are they exposed in the /sys filesystem? Is @TobleMiner willing to build 4 extra variants with +1/-1 for each of the 2 registers?

Reply:
@TobleMiner so from one new laptop with a reported 866.7Mbps AC link in both directions via br-lan and eth0 to another laptop with a Intel 82579LM Gigabit.


Slightly more variance in the numbers, but both directions above 200Mpbs.
Still not as good as I would have hoped though, even though the CPU utilisation is very low at the time (reported by iperf3 itself through the -V option)

```
PS D:\TEMP\iperf3.6_64bit\iperf3.6_64bit> .\iperf3.exe -c 192.168.1.124 -V
iperf 3.6
CYGWIN_NT-10.0 REM-VANDUERM 2.10.0(0.325/5/3) 2018-02-02 15:16 x86_64
warning: Ignoring nonsense TCP MSS -2146493952
Control connection MSS 0
Time: Sun, 13 Jan 2019 12:16:31 GMT
Connecting to host 192.168.1.124, port 5201
      Cookie: zruvz3ykl2bmesjcvd5nzfgn2lyhexf3b6xo
      TCP MSS: 0 (default)
[  5] local 192.168.1.170 port 23614 connected to 192.168.1.124 port 5201
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.01   sec  26.8 MBytes   223 Mbits/sec
[  5]   1.01-2.00   sec  29.9 MBytes   252 Mbits/sec
[  5]   2.00-3.00   sec  29.8 MBytes   249 Mbits/sec
[  5]   3.00-4.00   sec  30.8 MBytes   258 Mbits/sec
[  5]   4.00-5.01   sec  22.6 MBytes   189 Mbits/sec
[  5]   5.01-6.00   sec  17.0 MBytes   144 Mbits/sec
[  5]   6.00-7.00   sec  30.8 MBytes   257 Mbits/sec
[  5]   7.00-8.00   sec  30.2 MBytes   254 Mbits/sec
[  5]   8.00-9.01   sec  31.6 MBytes   264 Mbits/sec
[  5]   9.01-10.01  sec  21.4 MBytes   178 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.01  sec   271 MBytes   227 Mbits/sec                  sender
[  5]   0.00-10.01  sec   271 MBytes   227 Mbits/sec                  receiver
CPU Utilization: local/sender 32.9% (0.9%u/32.0%s), remote/receiver 5.7% (1.6%u/4.1%s)

iperf Done.
PS D:\TEMP\iperf3.6_64bit\iperf3.6_64bit> .\iperf3.exe -c 192.168.1.124 -V -R
iperf 3.6
CYGWIN_NT-10.0 REM-VANDUERM 2.10.0(0.325/5/3) 2018-02-02 15:16 x86_64
warning: Ignoring nonsense TCP MSS -2146493952
Control connection MSS 0
Time: Sun, 13 Jan 2019 12:16:45 GMT
Connecting to host 192.168.1.124, port 5201
Reverse mode, remote host 192.168.1.124 is sending
      Cookie: kdugkekygutidy2membg5ul4cakmepvqkkbo
      TCP MSS: 0 (default)
[  5] local 192.168.1.170 port 23618 connected to 192.168.1.124 port 5201
Starting Test: protocol: TCP, 1 streams, 131072 byte blocks, omitting 0 seconds, 10 second test, tos 0
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.01   sec  25.6 MBytes   212 Mbits/sec
[  5]   1.01-2.00   sec  25.1 MBytes   213 Mbits/sec
[  5]   2.00-3.00   sec  36.1 MBytes   303 Mbits/sec
[  5]   3.00-4.00   sec  37.4 MBytes   313 Mbits/sec
[  5]   4.00-5.00   sec  38.0 MBytes   319 Mbits/sec
[  5]   5.00-6.00   sec  30.9 MBytes   259 Mbits/sec
[  5]   6.00-7.01   sec  20.6 MBytes   172 Mbits/sec
[  5]   7.01-8.00   sec  36.5 MBytes   307 Mbits/sec
[  5]   8.00-9.00   sec  38.1 MBytes   319 Mbits/sec
[  5]   9.00-10.00  sec  37.1 MBytes   311 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
Test Complete. Summary Results:
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-10.00  sec   325 MBytes   273 Mbits/sec                  sender
[  5]   0.00-10.00  sec   325 MBytes   273 Mbits/sec                  receiver
CPU Utilization: local/receiver 13.8% (5.6%u/8.2%s), remote/sender 15.7% (4.5%u/11.1%s)

iperf Done.
PS D:\TEMP\iperf3.6_64bit\iperf3.6_64bit>
```

Reply:
That looks somewhat more like what I was expecting There are of  course a lot of variables that might be influencing wireless speed. On the NanoStation AC loco I'm breaking the 350 Mbit/s mark pretty much all the time. I'm testing against a ThinkPad T460s with an Intel 8260 wireless card.

Do you have VHT80 enabled on the NanoBeam? Please run htop on the NanoBeam while testing with iperf3 and configure htop to show both kernel threads and detailed CPU usage. Does ksoftirqd use pretty much all CPU?

Reply:
I guess so. CPU does go to 92%, ksoftirq is the top line, but only at 46%, so not quite most CPU.
But it is the biggest single user by far, yes.

Also, htop is not part of your builds. Had to do another ip & cable runaround to get it loaded.

![image](https://user-images.githubusercontent.com/3198100/51088302-367c6f00-175e-11e9-9b54-64027487b909.png)

Reply:
So, @TobleMiner, @ynezz ; I guess with that we are closing the eth0 config and delay value topic? @ynezz also Ok with the retrieved AirOS values? No testing of a build with new kernel?

I am asking, not because I do not want to invest more time, but if no more questions, I think I could clean up my desk and box/shelf everything up. It's getting a bit of a mess to start the new work week.

Reply:
@cybermaus Oops, sorry. I really though I included htop.

Thanks for posting the htop screenshot. It shows very clearly that the CPU is maxed out by softirqs (pink on the CPU meter). Thus the current bottleneck is probably just efficiency of the ath10k IRQ routine and its deferred work. 

Thanks for dumping the RGMII MAC config register, too. It confirms that even in the factory firmware both rxd-delay and rdv-delay are set to 3.

I think you can indeed put the device off to the side now since we do have all the hardware specific setting figured out. Thanks :smile: